TST: stats: free-threaded failure in `test_non_finite_inputs_and_int_bins`

### Describe your issue.

Running free-threading tests over and over until I see a failure, I see a test failure inside scipy.stats.tests.test_binned_statistic.TestBinnedStatistic.test_non_finite_inputs_and_int_bins.

It appears to be caused by concurrent modification of the self.u variable.

I observe this error on the free-threaded build, but I think it can happen on the non free threaded build too.

### Reproducing Code Example

This test command produces a test failure in about  1 / 10 attempts on my platform.

```bash
while OMP_NUM_THREADS=2 spin test -s stats -- -k test_binned_statistic -x --parallel-threads=10 --iterations=1 --skip-thread-unsafe=true; do sleep 1; done
```

The crash can be made more consistent by adding the following random sleep, and upping `--iterations` to 10.

```diff
index ea5ff1d053..a91c76ee18 100644
--- a/scipy/stats/tests/test_binned_statistic.py
+++ b/scipy/stats/tests/test_binned_statistic.py
@@ -64,17 +64,18 @@ class TestBinnedStatistic:
         # if either `values` or `sample` contain np.inf or np.nan throw
         # see issue gh-9010 for more
         x = self.x
         u = self.u
         orig = u[0]
         u[0] = np.inf
+        import random
+        import time
+        if random.random() < 0.5:
+            time.sleep(0.1)
         assert_raises(ValueError, binned_statistic, u, x, 'std', bins=10)
         # need to test for non-python specific ints, e.g. np.int8, np.int64
         assert_raises(ValueError, binned_statistic, u, x, 'std',
                       bins=np.int64(10))
         u[0] = np.nan
         assert_raises(ValueError, binned_statistic, u, x, 'count', bins=10)
         # replace original value, u belongs the class
         u[0] = orig
 
     def test_1d_result_attributes(self):
         x = self.x
```

### Error message

```shell
··································································································· [ 86%]
·························································ss········································ [ 86%]
···········sss····················································································· [ 86%]
···························································································         [ 86%]
scipy/stats/tests/test_binned_statistic.py ····e

================================================= ERRORS ==================================================
________________ ERROR at call of TestBinnedStatistic.test_non_finite_inputs_and_int_bins _________________

self = <scipy.stats.tests.test_binned_statistic.TestBinnedStatistic object at 0x5e3a35bc6d0>

    def test_non_finite_inputs_and_int_bins(self):
        # if either `values` or `sample` contain np.inf or np.nan throw
        # see issue gh-9010 for more
        x = self.x
        u = self.u
        orig = u[0]
        u[0] = np.inf
        assert_raises(ValueError, binned_statistic, u, x, 'std', bins=10)
        # need to test for non-python specific ints, e.g. np.int8, np.int64
>       assert_raises(ValueError, binned_statistic, u, x, 'std',
                      bins=np.int64(10))
E       Failed: DID NOT RAISE <class 'ValueError'>

orig       = np.float64(nan)
self       = <scipy.stats.tests.test_binned_statistic.TestBinnedStatistic object at 0x5e3a35bc6d0>
u          = array([1000000.69958655, 1000000.65100148, 1000000.41635223,
       1000000.98462205, 1000000.79918981, 1000000.625309...0000.35232519, 1000000.40770789,
       1000000.01995802, 1000000.8749159 , 1000000.59799903,
       1000000.12496151])
x          = array([0.78050106, 0.37008054, 0.13692841, 0.71948323, 0.62578559,
       0.63937671, 0.7835862 , 0.8034948 , 0.940272...01, 0.20262148, 0.39515713, 0.84992089, 0.54980338,
       0.71266933, 0.24569699, 0.53163105, 0.58695467, 0.03554972])

scipy/stats/tests/test_binned_statistic.py:72: Failed
```

### SciPy/NumPy/Python version and system information

```shell
1.17.0.dev0+git20251031.f8f9031 2.3.4 sys.version_info(major=3, minor=14, micro=0, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /home/nodell/scipy-ft/venv/lib/python3.14t/site-packages/scipy_openblas32/include
    lib directory: /home/nodell/scipy-ft/venv/lib/python3.14t/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.30 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /home/nodell/scipy-ft
    version: 0.3.30
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /home/nodell/scipy-ft/venv/lib/python3.14t/site-packages/scipy_openblas32/include
    lib directory: /home/nodell/scipy-ft/venv/lib/python3.14t/site-packages/scipy_openblas32/lib
    name: scipy-openblas
    openblas configuration: OpenBLAS 0.3.30 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64
    pc file directory: /home/nodell/scipy-ft
    version: 0.3.30
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 3.0.1
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 14.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 14.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.1.6
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 14.2.1
  pythran:
    include directory: ../venv/lib/python3.14t/site-packages/pythran
    version: 0.18.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /home/nodell/scipy-ft/venv/bin/python
  version: '3.14'
```

--- a/scipy/stats/tests/test_binned_statistic.py
+++ b/scipy/stats/tests/test_binned_statistic.py
@@ -64,8 +64,7 @@ class TestBinnedStatistic:
         # if either `values` or `sample` contain np.inf or np.nan throw
         # see issue gh-9010 for more
         x = self.x
-        u = self.u
-        orig = u[0]
+        u = self.u.copy()  # take copy before modification
         u[0] = np.inf
         assert_raises(ValueError, binned_statistic, u, x, 'std', bins=10)
         # need to test for non-python specific ints, e.g. np.int8, np.int64
@@ -73,8 +72,6 @@ class TestBinnedStatistic:
                       bins=np.int64(10))
         u[0] = np.nan
         assert_raises(ValueError, binned_statistic, u, x, 'count', bins=10)
-        # replace original value, u belongs the class
-        u[0] = orig
 
     def test_1d_result_attributes(self):
         x = self.x

